Replace the 'unmediated' bool with a 'mediation' enum.

mikewest opened the pull request:
This seems sane to me based on our discussion, @jyasskin. It might also resolve one of @leshi's concerns about the current shape of the API.

/cc @battre

battre commented on index.src.html:
```diff
@@ -396,19 +396,25 @@
+  <dl dfn-for="CredentialRequestOptions" dfn-type="dict-member">
+    :   <dfn>mediation</dfn>
+    ::  This property specifies the mediation requirements for a given credential request. The
+        meaning of each enum value are described below in {{CredentialMediationRequirement}}.
```
nit: "meaning of each enum value **is**" or "**meanings** of **the enum values** are"

battre commented on index.src.html:
```diff
@@ -396,19 +396,25 @@
 
+  <div class="note">
+    Earlier versions of this specification defined a boolean `unmediated` member. Setting that
+    to `true` had similar effect to setting {{CredentialRequestOptions/mediation}} to "silent", and
```
nit: had *a* similar?

battre commented on index.src.html:
```diff
@@ -459,6 +465,48 @@
+  </pre>
+  <dl dfn-for="CredentialMediationRequirement" dfn-type="enum-value">
+    :   <dfn>silent</dfn>
+    ::  User mediation is suppressed for the given operation. If it can be performed without user
```
I think the "it" is not well defined.

battre commented on index.src.html:
```diff
@@ -459,6 +465,48 @@
+
+    :   <dfn>optional</dfn>
+    ::  If credentials can be handed over without user mediation, they will be. If no such
+        credentials are available, then the user agent will ask the user for permission.
```
If credentials are available for the security origin but require user mediation (see [=requires user mediation=] flag), the user agent will ask the user for permission. ?

battre commented on index.src.html:
```diff
@@ -459,6 +465,48 @@
+    :   <dfn>optional</dfn>
+    ::  If credentials can be handed over without user mediation, they will be. If no such
+        credentials are available, then the user agent will ask the user for permission.
+
```
Note: The intended usage is to support a sign-in button, where showing a credential picker is in context and would not surprise the user.

mikewest commented:
Thanks, @battre: I've addressed your feedback in https://github.com/w3c/webappsec-credential-management/pull/73/commits/49943df289333bc4bb7b30a39dddc5a425d772a4.

battre commented on index.src.html:
```diff
@@ -459,6 +465,56 @@
+    :   <dfn>silent</dfn>
+    ::  User mediation is suppressed for the given operation. If the operation can be performed
+        without user involvement, wonderful. If user involvement is necessary, then the operation
+        will fail rather than involving the user.
```
I think it would not fail but resolve to `null`, right?

battre commented on index.src.html:
```diff
@@ -459,6 +465,56 @@
+        Note: For some operations, developers may wish to ensure that the user is involved, even if
+        they've otherwise chosen to stay signed-in. For example, a website might support sign-in
+        with multiple accounts at once, and it would be unfortunate if {{get()}} automatically
+        returned credentials without asking for disambiguation. In these cases, the developer can
```
As discussed, I think it would be nice to make this scenario clearer.

Another point is that some password managers / authenticators may require a master password / biometric reauth as part of the mediation, so setting required to true would make a security difference.

mikewest commented:
Thanks! I took another pass in  https://github.com/w3c/webappsec-credential-management/pull/73/commits/5dba302bd8b7fb260c9376bda7b56877d12d0f97.

battre commented on index.src.html:
```diff
@@ -459,6 +465,119 @@
+  <div class="example" id="example-mediation-switch">
+    MegaCorp, Inc. wishes to support switching from one user account to another. In order to ensure
+    that credentials aren't returned automatically in response to clicking on a "Change user" or "Add
+    account" button, , MegaCorp, Inc. can call {{get()}} with a
```
nit extra ", "

mikewest replied:
Good eye!

battre commented on index.src.html:
```diff
@@ -459,6 +465,119 @@
+      });
+    </pre>
+  </div>
+
```
Should we another example here "example-mediation-optional" for the sake of completeness?

MegaCorp, Inc. does not want to use silent sign-in but wishes to make the sign-in button on their homepage as convenient as possible. If exactly one credential is available and not mediation is required, that credential should be returned silently. Otherwise, the user can be prompted to pick a credential. --> optional?

mikewest replied:
Sure. I can add something like that.

mikewest replied:
Added.

battre commented on index.src.html:
```diff
@@ -459,6 +465,119 @@
+    before taking action. Even if a user has [[#user-mediation-requirement|opted-into]] signing in
+    without [=user mediation=], MegaCorp, Inc. can ensure that the user agent requires mediation
+    by calling {{get()}} with a {{CredentialRequestOptions/mediation}} member set to
+    "{{CredentialMediationRequirement/required}}":
```
add something like "Depending on the security model of the browser or the credential type, this may require entering a master password or a biometric reauthentication before a credential is passed to the website."?

mikewest replied:
Hrm. I don't follow. If an authenticator requires a thumbprint, isn't the credential already unavailable for unmediated access? In that case, `required` is already more or less a no-op.

battre replied:
I was thinking of a hypothetical authenticator that can require a fingerprint but does not have to.

For example your phone: Assume you have exactly one credential on your phone. The phone has a fingerprint sensor. Now the phone could be configured to only release a password if (a) usermediation is off, or (b) you swipe your finger.

mikewest replied:
Added a note along the lines of what @battre asked for.

battre commented on index.src.html:
```diff
@@ -459,6 +465,119 @@
+    </pre>
+  </div>
+
+  <div class="example" id="example-mediation-switch">
```
I think that this description is not very convincing, yet. The usecase could be implemented by a call to requireUserMediation() and get({..., mediation: "optional"}), right?

mikewest replied:
That would also have effect on every subsequent call to `get()`. That is, we don't sign you out of Google if you visit `passwords.google.com` and we ask for reauthentication, right? It's a case-by-case basis.

jyasskin commented on index.src.html:
```diff
@@ -459,6 +465,140 @@
+  </div>
+
+  <div class="example" id="example-mediation-require">
+    MegaCorp, Inc. wishes to protect a sensitive operation by requiring a user to reauthenticate
```
Using the same value for reauthentication and "add identity" makes me worried. For "add identity", we'd generally want a pretty smooth flow where the user just selects the new identity, but for reauthentication, there's a good argument for the UA to double-check that it's the right person. Even if Chrome wouldn't double-check, I'd like to preserve UA freedom to do so.

However, I also don't want to get into another `autocomplete=off` situation where sites force user-verification in hostile ways.

It'd be nice to get some feedback from site operators about what behaviors they expect and/or dislike, before this change ships. However, I think this change is a clear improvement over what's currently in the spec, and I'm happy with merging before getting that feedback.

mikewest replied:
Let's land this and continue that conversation in the next patch. Would you mind filing an issue with your concerns?

jyasskin commented on index.src.html:
```diff
@@ -459,6 +465,140 @@
+
+  <div class="example" id="example-mediation-switch">
+    MegaCorp, Inc. wishes to support switching from one user account to another. In order to ensure
+    that credentials aren't returned automatically in response to clicking on a "Change user" or "Add
```
Maybe, "... ensure that the user gets a chance to select a different identity ...", to express the site's goal instead of one of many possible anti-goals?

mikewest replied:
Rephrased along these lines.

jyasskin commented on index.src.html:
```diff
@@ -459,6 +465,140 @@
+  </div>
+
+  <div class="example" id="example-mediation-switch">
+    MegaCorp, Inc. wishes to support switching from one user account to another. In order to ensure
```
"Switching from one user account to another" could be accomplished by logging out of the first, but so it might be clearer to say they wish to support logging into two accounts/identities at once.

mikewest replied:
Indeed.

mikewest commented on index.src.html:
```diff
@@ -396,19 +396,25 @@
     [Exposed=Window, SecureContext]
     dictionary CredentialRequestOptions {
-      boolean unmediated = false;
+      CredentialMediationRequirement mediation = "allowed";
```
Thanks!

mikewest commented on index.src.html:
```diff
@@ -459,6 +465,140 @@
+        Note: This requirement is intended to support [reauthentication](#example-mediation-require)
+        or [user-switching](#example-mediation-switch) scenarios. Further, the requirement is tied
+        to a specific operation, and does not affect the [=requires user mediation=] flag for the
+        origin. To toggle that flag, developers should call {{requireUserMediation()}}.
```
Indeed.
